compiler/rustc_lint/src/unused.rs

```diff
@@ -1333,7 +1334,11 @@ impl EarlyLintPass for UnusedParens {
                 self.with_self_ty_parens = false;
             }
             ast::TyKind::Ref(_, mut_ty) | ast::TyKind::Ptr(mut_ty) => {
-                self.in_no_bounds_pos.insert(mut_ty.ty.id, NoBoundsException::OneBound);
+                // If this type itself appears in no-bounds position, we propagate its
+                // potentially tighter constraint or risk a false posive (issue 143653).
+                let own_constraint = self.in_no_bounds_pos.get(&ty.id).copied();
+                let constraint = own_constraint.unwrap_or(NoBoundsException::OneBound);
+                self.in_no_bounds_pos.insert(mut_ty.ty.id, constraint);
```

Comment:
in a sense we want tis to be `constraint` = `own_constraint.max(NoBoundsException::OneBound)` where as `NoBoundsException::OneBound` is currently the minimal value this is correct?

I would feel safer if we were to make this a match on `own_constraint` or add a `NoBoundsException.merge` function or sth :thinking: